### PR TITLE
Initial commit of Java style guide.

### PR DESCRIPTION
**Pull request by benkomalo**
This is mostly intended for Java on Android, since that's the only
foreseeable use of Java at Khan Academy.

Non-urgent request, given we're still in the planning phases, but @NachoSoto @alangpierce @mgp or anyone else have opinions or additions they'd like to suggest here?

### CODE REVIEW

**Comment by NachoSoto** [comment]
Awesome! thank you for getting this started.

I would strongly suggest that we make the style guide be the Android Studio code style XML configuration file. I'm pretty flexible when it comes to styles (with some exceptions), so long as we are consistent AND we spend close to no time *linting* code as part of code review. With that in mind, it would be great if we all formatted code automatically using the same configuration. I haven't looked much into it, but they seem to live in `~/Library/Preferences/AndroidStudio/codestyles` and I think there's a way to commit them with your project :)

File: style/java.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# Java
+
+At Khan Academy, we only use Java for Android and, as such, follow the [Android Code Style Guidelines for Contributors](https://source.android.com/source/code-style.html). Note that those rules themselves extend the [standard Java language style guide as of 1999](http://www.oracle.com/technetwork/java/javase/documentation/codeconventions-141855.html). The rest of this document describes additions and clarifications to those rules that we follow at Khan Academy.
```
Comment (alangpierce) [inline style/java.md]: > we only use Java for Android

One minor exception that I can think of:
https://github.com/Khan/ka-pycharm-plugin/blob/master/src/org/khanacademy/pycharmplugin/KATestConfigurationProducer.java

Not a whole lot there, but that's what my Java looks like normally.

File: style/java.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+
+## Imports
+
+Import statements are divided into the following groups, in this order, with each group separated by a blank line:
```
Comment (alangpierce) [inline style/java.md]: I would prefer to just do whatever Android Studio's default algorithm is, and not even document the order unless we expect people to be working on the project outside of Android Studio. But if we use a shared set of Android Studio settings and they encode the rules below, then I guess I would be fine with that as well. IMO, Java imports can and should be seen as "metadata for the compiler", and Java programmers should never have to read, write, or think about imports, except in the rare case when there are two classes with the same name.

But as long as a new person joining the project doesn't have to learn the import rules and doesn't have to do any additional setup just to get "organize imports" working, I'm fine with whatever we end up using.

Comment (NachoSoto) [inline style/java.md]: :100:

Comment (benkomalo) [inline style/java.md]: Right - I was thinking that we'd do both codify and export the XML file to enforce it. You can also check it in with the project, so by default people cloning the main Android project would adopt it as well. But if we were to have multiple projects, anyone starting a new one would either have to know to import the XML to their global/personal settings or seed it or whatever.

**Comment by alangpierce** [comment]
Cool, I'm happy with the Android style guide overall, and in almost every case it's the style that I would be using anyway.

A few places where we should think about diverging:
* I'm not a fan of the `m` and `s` prefixes for non-static and static fields. Google's usual Java style ( https://google-styleguide.googlecode.com/svn/trunk/javaguide.html ) doesn't use those, and I can't think of any other Java projects that I've seen that use them. Following Android's convention has its upsides, but I'm also a little worried about the style bleeding into more general-purpose Java libraries that we might write. Plus I think it just looks a little nicer to not use hungarian notation.
* I don't think we should require javadoc for all public methods.
* TODO comments should use our existing Python/JS style, obviously.

Other possible points of contention that I think we should think about before writing too much code:
* I think we should use `@Override` everywhere that it's allowed, including on methods that satisfy an interface. I like the ability to see `@Override public` and have it mean "super-public".
* I think normal fields should always go at the top of classes, not next to methods. I have less of a preference for constants, though.
* I prefer to avoid same-line `if` statements.
* I think classes should have exactly one constructor, and it should do nothing except assign field values and possibly validate invariants.
* Should we make a matching interface for every business logic class?
* What naming convention do we want for related classes and interfaces? I prefer `Foo` for the interface and `FooImpl` for the class, but another approach is `IFoo` for the interface and `Foo` for the class.
* For "data object" classes, do we want to always use getters, or is it ok to just use fields directly?

I imagine there's plenty more to argue about, but those are my initial thoughts.

**Comment by benkomalo** [comment]
Thanks for looking @alangpierce! Responses inline, and I'll make changes tomorrow....

> I'm not a fan of the `m` and `s` prefixes for non-static and static fields

I was put off by it initially, as well, but didn't actually mind if after I got used to it, and found it to be helpful in the end. I was also just inclined to stick with it because realistically we'd be doing mostly Android Java. I hear your concern about it "leaking to non-Android" projects, but I don't feel like that would be a big loss, tbh. On the flipside, there isn't a strong need to even be consistent with Android internals, since we're unlikely to refer to those fields from an app. So I guess I could go either way here. If other people feel this way, I'm happy to drop it.

> I don't think we should require javadoc for all public methods.
> TODO comments should use our existing Python/JS style, obviously.

Agree with both. I'll amend as such.

> I think we should use `@Override` everywhere that it's allowed,

Agreed. I thought the Android style guide already included that, but I guess it only says for overriding a superclass method. Maybe it's just old and written pre 1.6? I'll add this, too, in any case.

> I think normal fields should always go at the top of classes,
> I prefer to avoid same-line if statements.

Agreed on both.

> I think classes should have exactly one constructor

Hrm. This is a bit heavier of a requirement. It's also not going to be apply to many things in Android, as you'll likely be making a lot of `View` or `Activity` or `Fragment` or whatever subclasses, and will, as such, require certain constructors.

> Should we make a matching interface for every business logic class?

Also seems a bit heavy to me, though it's worth keeping an eye out for it.

> What naming convention do we want for related classes and interfaces?

I'm used to `Foo` and `FooImpl` as well, though in Android I actually see very little of either prefix/suffixing. e.g. `ListAdapter` as interface and `BaseAdapter` as a class are standard Android-y names. My vote is to hold off on codifying this in the style guide for now?

> For "data object" classes, do we want to always use getters, or is it ok to just use fields directly?

Hrm. Could go either? I don't feel strong enough about this to want to codify this either. I can see situations where I'd want one and the other, though most of the time I'd probably default to getters.

**Comment by alangpierce** [comment]
Yeah, I've worked comfortably in codebases with and without hungarian notation, and I think neither approach is obviously better than the other. Consistency within our code is way more important, obviously. If I had to choose, though, I think I would still avoid hungarian notation for new code, and the internet seems to mostly agree with me.

For the "one constructor" rule, I agree that there may be specialized cases where you may need to define multiple constructors so you can call different types of `super()` when subclassing SDK code. I didn't mean that "one constructor" should be a lint rule or anything like that; more than it should be a best practice.

More generally, I think the Android SDK goes against some things that I would consider best practices. For example, it heavily uses inheritance (when it could use composition), it expects application code to use non-final fields almost everywhere, and it exposes a modify-after-creation approach for configuration (instead of using builders). I think that it's fine for us to write code in that style to work with the SDK, but I would rather keep things immutable and and composable in the areas of the code that we have more control over.

To be clear, I'm not saying that there should only be one way to instantiate a class, I'm saying that any non-trivial way to instantiate a class should be exposed through static factory methods that use a shared simple (which usually means IDE-generated) constructor. I think there are about 10 reasons why static factory methods are better than constructors, but this comment is getting too long to go through all of those :-) .

Anyway, I guess some of this is getting into "best practices" rather than specific "style guide" rules, and even the naming-related things can be figured out as we go. Just things to keep in mind for whenever they come up.

**Comment by benkomalo** [comment]
Awesome. Thanks again for the feedback @NachoSoto and @alangpierce 

I've pushed some amendments to include some additional stuff Alan suggested. 

The two semi outstanding issues in comments here:

1. The import rule also requires IntelliJ config files to enforce - I will check this into the project at the least but will also own trying to get it into onboarding docs. Totally agree the tooling support is just as important as the doc though. I don't think we should "stick with the default IDE config" because the default one doesn't separate out `org.khanacademy.*` imports and at the least, I would want those out in its own group.

2. There are a lot of "best practices" Alan's pointed out here in that response. I agree with them generally, though am unsure about it belonging in a style guide. I think a style guide should encourage best practices where possible, but sometimes it's difficult to formulate unambiguously/succinctly. Let's keep those ones out of the initial commit unless anyone feels strongly? Though am happy to keep discussing/being mindful of them and put in something in a later revision if we can figure out a way to formulate a style rule for them (that's also mindful of the fact that practically speaking we might be working amidst Android code that will break some of these best practices).

Other than those two, I *think* this is good to go. So I'll land later this afternoon unless anyone else has any objections / followups on above.

**Comment by alangpierce** [comment]
LGTM. I'm not convinced it's that useful to separate first-party from third-party imports, but I don't object to it either.

**Comment by NachoSoto** [comment]
Agreed on all fronts!

About multiple constructors: I would actually suggest that we make it a hard rule to have **only** one *designated initializer*: i.e. there can be multiple *convenience* initializers, but all of them have to use the designated. This is the pattern used in Cocoa, I'll be happy to explain the benefits if you want me to :)

**Pull request merged**